pkg/services/datasources/service/secret_migration_service.go

```diff
@@ -0,0 +1,76 @@
+package service
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/models"
+	"github.com/grafana/grafana/pkg/services/datasources"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/secrets/kvstore"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+type SecretMigrationService struct {
+	dataSourcesService datasources.DataSourceService
+	secretsStore       kvstore.SecretsKVStore
+	features           featuremgmt.FeatureToggles
+	log                log.Logger
+}
+
+func ProvideSecretMigrationService(
+	cfg *setting.Cfg, dataSourcesService datasources.DataSourceService,
+	secretsStore kvstore.SecretsKVStore, features featuremgmt.FeatureToggles,
+) datasources.SecretMigrationService {
+	return &SecretMigrationService{
+		dataSourcesService: dataSourcesService,
+		secretsStore:       secretsStore,
+		features:           features,
+		log:                log.New("secret.migration"),
+	}
+}
+
+func (s *SecretMigrationService) Run(ctx context.Context) error {
+	query := &models.GetDataSourcesQuery{}
+	err := s.dataSourcesService.GetDataSources(ctx, query)
+	if err != nil {
+		return err
+	}
+
+	s.log.Debug("starting data source secret migration")
+	for _, ds := range query.Result {
+		hasMigration, _ := ds.JsonData.Get("secretMigrationComplete").Bool()
+		if !hasMigration {
+			secureJsonData, err := s.dataSourcesService.DecryptLegacySecrets(ctx, ds)
+			if err != nil {
+				return err
+			}
+
+			jsonData, err := json.Marshal(secureJsonData)
+			if err != nil {
+				return err
+			}
+
+			err = s.secretsStore.Set(ctx, ds.OrgId, ds.Name, secretType, string(jsonData))
+			if err != nil {
+				return err
+			}
+
+			ds.JsonData.Set("secretMigrationComplete", true)
```

Comment:
Is there anywhere else we could put this flag? Probably not a big deal if this does make it to the client, but I would say it is probably better if it doesn't.

Reply:
We could probably match the keys in the secrets table with the data source table.

Reply:
Using the keys would result in O(n*m) complexity, I think using the JsonData is the best way for now.